src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java

```diff
@@ -0,0 +1,94 @@
+package uk.gov.companieshouse.accounts.association.tasks;
+
+import java.util.stream.Collectors;
+import org.springframework.data.mongodb.core.query.Update;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+import uk.gov.companieshouse.accounts.association.models.AssociationDao;
+import uk.gov.companieshouse.accounts.association.service.AssociationsService;
+import uk.gov.companieshouse.accounts.association.service.UsersService;
+import uk.gov.companieshouse.accounts.association.utils.StaticPropertyUtil;
+import uk.gov.companieshouse.logging.Logger;
+import uk.gov.companieshouse.logging.LoggerFactory;
+
+@Component
+public class DataMigrationPostProcessingTask {
+
+    private final AssociationsService associationsService;
+
+    private final UsersService usersService;
+
+    private static final Logger LOG = LoggerFactory.getLogger( StaticPropertyUtil.APPLICATION_NAMESPACE );
+
+    private static final String DATA_MIGRATION_USER_ID_UPDATE_TASK = "data-migration-user-id-update-task";
+
+    // TODO: choose different value e.g. 15
+    private static final int ITEMS_PER_PAGE = 2;
+
+    public DataMigrationPostProcessingTask( final AssociationsService associationsService, final UsersService usersService ) {
+        this.associationsService = associationsService;
+        this.usersService = usersService;
+    }
+
+    // TODO: change cron = "0 0 0 * * ?", so that the cron job runs everyday at midnight
+    @Scheduled( cron = "0 2 17 * * ?" )
+    public void processMigratedAssociations(){
+        LOG.infoContext(DATA_MIGRATION_USER_ID_UPDATE_TASK, "Starting up data migration user_id task...", null );
+
+        final var totalAssociations = associationsService.fetchNumberOfUnprocessedMigratedAssociations();
+        final var totalPages = ( totalAssociations / ITEMS_PER_PAGE ) + ( totalAssociations % 2 == 0 ? 0 : 1 );
+        int currentPage = (int) totalPages - 1;
+
+        LOG.debugContext(DATA_MIGRATION_USER_ID_UPDATE_TASK, String.format( "Identified %s associations to be processed in %s batches in reverse order... starting with batch %s", totalAssociations, totalPages, currentPage ), null );
+
+        while ( currentPage > -1 ) {
+            final var page = associationsService.fetchUnprocessedMigratedAssociations( currentPage, ITEMS_PER_PAGE );
+
+            final var users = usersService.searchUserDetails( page.stream() );
+
+            page.stream()
+                    .filter( associationDao -> users.containsKey( associationDao.getUserEmail() ) )
```

Comment:
if page already has associations where email exists. why are we filtering again?

Reply:
You are correct - the page only contains associations where the email exists.

On line 47, we extract the `user_email`s from the associations, and pass it to the accounts-user-api to fetch the users (because we need the 'user_id'). Because the user might, or might not have a CHS account, the accounts-user-api might or might not actually find all of the users. So the `users` variable is a subset of the users from across the associations, i.e. all of the ones that actually have a `user_id`.

On line 50, the filter extracts the 'user_email' from the association, and checks if it exists in 'users' - if it does then this means the user in the association has a 'user_id'.

In other words, this is not the same filter as the one in the query.